operators/pkg/tenant-controller/sandbox.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020-2021 Politecnico di Torino
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tenant_controller groups the functionalities related to the Tenant controller.
+package tenant_controller
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	crownlabsv1alpha2 "github.com/netgroup-polito/CrownLabs/operators/api/v1alpha2"
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/forge"
+)
+
+func (r *TenantReconciler) enforceSandboxResources(ctx context.Context, key types.NamespacedName, tn *crownlabsv1alpha2.Tenant, sbnsName string) (nsOk bool, err error) {
+	log := ctrl.LoggerFrom(ctx, "sandbox", sbnsName) // retrieve log from context
+	ns := v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: sbnsName}}
+	labels, updated := forge.SandboxLabels(ns.Labels, tn.Name, r.TargetLabelKey, r.TargetLabelValue)
+
+	if updated {
+		original := tn.DeepCopy()
+		tn.SetLabels(labels)
+		if err := r.Patch(ctx, tn, client.MergeFrom(original)); err != nil {
+			log.Error(err, "failed to update the sandbox labels")
+			return false, err
+		}
+		log.Info("sandbox labels correctly configured")
+	} else {
+		return false, nil
+	}
+
+	if _, nsErr := ctrl.CreateOrUpdate(ctx, r.Client, &ns, func() error {
+		return ctrl.SetControllerReference(tn, &ns, r.Scheme)
+	}); nsErr != nil {
+		log.Error(nsErr, "failed to update sandbox namespace", "tenant", tn.Name)
+		return true, nsErr
+	}
+	log.Info("sandbox namespace correctly updated", "tenant", tn.Name)
```

Comment:
```suggestion
	res, err := ctrl.CreateOrUpdate(ctx, r.Client, &namespace, func() error {
          	namespace.SetLabels(forge.TenantObjectLabels(namespace.GetLabels(), tenant))
		return ctrl.SetControllerReference(tn, &ns, r.Scheme)
	})
	if err != nil {
		log.Error(err, "failed to enforce object", "namespace", klog.KObj(&namespace))
		return err
	}
	log.V(utils.FromResult(res)).Info("object enforced", "namespace", klog.KObj(&namespace), "result", res)
```

Please, do the same for the other objects below.